[Merged by Bors] - bevy_render: Support removal of nodes, edges, subgraphs

superdump opened the pull request:
Add support for removing nodes, edges, and subgraphs. This enables live re-wiring of the render graph.

This was something I did to support the MSAA implementation, but it turned out to be unnecessary there. However, it is still useful so here it is in its own PR.

dataphract commented on pipelined/bevy_render2/src/render_graph/graph.rs:
```diff
@@ -176,8 +228,39 @@ impl RenderGraph {
+        Ok(())
+    }
+
+    pub fn validate_edge(
+        &mut self,
+        edge: &Edge,
+        should_exist: bool,
+    ) -> Result<(), RenderGraphError> {
```
The `bool` argument here is tough to grok at the call site:

```rust
validate_edge(&edge, false);
```

It's not clear what `false` means unless you go look at the docs and read the argument name. (See [C-CUSTOM-TYPE](https://rust-lang.github.io/api-guidelines/type-safety.html#arguments-convey-meaning-through-types-not-bool-or-option-c-custom-type) in the Rust API guidelines.)

Something like

```rust
enum EdgeExists {
    Yes, // Or `True`, or `Exists`
    No,  // Or `False`, or `DoesNotExist`
}
```

would be more explicit.

superdump replied:
Done.

jakobhellermann commented on pipelined/bevy_render2/src/render_graph/graph.rs:
```diff
@@ -58,6 +58,16 @@ impl RenderGraph {
         id
     }
 
+    pub fn remove_node(&mut self, name: impl Into<Cow<'static, str>>) {
```
`remove_node` will leave edges that referenced the node around if there were any. For `A -> B` `graph.remove("A")` you'd need to remove the `A` node, and remove all `input_edges` (I think) of `B` that reference `A`.

superdump replied:
Adding a slot edge or node edge from an output to an input creates an Edge of the appropriate type and adds it to both the output node's output edges and the input node's input edges. As such, I think what needs to be done is to look at all the input edges for the node, and remove that edge from the edge's output node's output edges, and look at all the output edges for the node, and remove that edge from the edge's input node's input edges. I made this change.

superdump commented:
This is now updated on top of `main`. I've also addressed the feedback to get it into a better state. However, I haven't tested it other than to check it builds and passes `cargo run -p ci`. I suppose it should have a bunch of unit tests.

superdump commented:
Also, when I changed the PR base, it looks like CI broke?

superdump commented:
bors try

jakobhellermann commented on crates/bevy_render/src/render_graph/graph.rs:
```diff
@@ -101,6 +103,70 @@ impl RenderGraph {
+                                output_node
+                                    .edges
+                                    .remove_output_edge(input_edge.clone())
+                                    .ok();
```
The `.ok()` here doesn't do anything, is the intention of it to silence the "unused output" warning or is it a leftover from a previous version?

jakobhellermann replied:
maybe this method could return a `Result<(), RenderGraphError>` like the other methods

superdump replied:
Done.

jakobhellermann commented on crates/bevy_render/src/render_graph/graph.rs:
```diff
@@ -101,6 +103,70 @@ impl RenderGraph {
+                for input_edge in node_state.edges.input_edges.iter() {
+                    match input_edge {
+                        Edge::SlotEdge {
+                            input_node: _,
+                            input_index: _,
+                            output_node,
+                            output_index: _,
```
```suggestion
                            output_node,
                            output_index: _,
                            input_node: _,
                            input_index: _,
```
minor change, but I find it much easier to read because it makes it clear that the edge goes from the output node to the input node (same at the other locations)

inodentry replied:
Agree with this. In some places, it is already the other way around. Should be `output` followed by `input` everywhere, for consistency.

superdump replied:
I really feel we should rename all of this because input and output are confusing. But sure, I can order them that way.

jakobhellermann reviewed (approved):
looks good, I left some small suggestions but other than that it should be good to merge

inodentry reviewed (approved):
LGTM. I got no more feedback to add, besides the minor things from the previous review.

cart commented on crates/bevy_render/src/render_graph/graph.rs:
```diff
@@ -99,6 +101,69 @@ impl RenderGraph {
+                            input_node,
+                            input_index: _,
+                        } => {
+                            if let Ok(input_node) = self.get_node_state_mut(*input_node) {
```
This relies on each edge to be properly represented as both an "input edge" and an "output edge", otherwise this could "early exit" without removing every input edge (same goes for the previous section). 

We can and should rely on the "bi-directional edge" behavior, but currently we can't because we expose public mutable access to `NodeState::edges::input_slots` and `NodeState::edges::output_slots`. This allows users to manually violate the guarantees we rely on. This was an oversight on my part when I implemented the graph. We should make these fields private.

cart reviewed (commented):
Generally looks good to me! We should clean up one "legacy correctness issue" that existed prior to these changes. I'll do that now so we can merge this asap.

cart commented:
bors r+
